File: problems/Climbing Stairs/climbStairs.cs
public class Solution {
    public int ClimbStairs(int n) {
        return climbStairsConstantSpace(n);
    }

    public int climbStairsConstantSpace(int n) {
        var minusOne = 1;
        var minusTwo = 1;

        for (var i = 2; n >= i; ++i) {
            var curr = minusOne + minusTwo;
            minusTwo = minusOne;
            minusOne = curr;
        }

        return minusOne;
    }

    public int climbStairsBottomUp(int n) {
        int[] dp = new int[1 + n];

        Array.Fill(dp, 1);

        for (int idx = 2; n >= idx; ++idx) {
            dp[idx] = dp[idx - 1] + dp[idx - 2];
        }


        return dp[n];
    }
}

File: problems/Climbing Stairs/climbStairs.java
class Solution {
    public int climbStairs(int n) {
        var dp = new int[1 + n];

        return countStairs(dp, n);
    }

    private int countStairs(int[] dp, int n) {
        if (0 == n || 1 == n) {
            return 1;
        }

        if (0 == dp[n]) {
            dp[n] = countStairs(dp, n - 1) + countStairs(dp, n - 2);
        }

        return dp[n];
    }
}
